public/src/admin/manage/categories.js

```diff
@@ -78,5 +44,156 @@ define('admin/manage/categories', function() {
 		});
 	};
 
+	Categories.render = function(categories){
+        var container = $('.categories');
+
+        if(!categories || categories.length == 0){
+            $('<div></div>')
+                .addClass('alert alert-info text-center')
+                .text('You have no active categories.')
+                .appendTo(container);
+        }else{
+            sortables = {};
+            renderList(categories, 0, container, 0);
+        }
+	};
+
+	function itemDidAdd(e){
+        newCategoryId = e.to.dataset.cid;
+	}
+
+	function itemDragDidEnd(e){
+        var isCategoryUpdate = (newCategoryId != -1);
+		//Update needed?
+		if((e.newIndex != undefined && e.oldIndex != e.newIndex) || isCategoryUpdate){
+            var parentCategory = isCategoryUpdate ? sortables[newCategoryId] : sortables[e.from.dataset.cid],
+                modified = {}, i = 0, list = parentCategory.toArray(), len = list.length;
+
+            for(i; i < len; ++i) {
+                modified[list[i]] = {
+                    order: (i + 1)
+                }
+            }
+
+            if(isCategoryUpdate){
+                modified[e.item.dataset.cid]['parentCid'] = newCategoryId;
+            }
+
+            newCategoryId = -1
+            socket.emit('admin.categories.update', modified);
+        }
+	}
+
+    /**
+     * Render categories - recursively
+     *
+     * @param categories {array} categories tree
+     * @param level {number} current sub-level of rendering
+     * @param container {object} parent jquery element for the list
+     * @param parentId {number} parent category identifier
+     */
+    function renderList(categories, level, container, parentId){
+        var i = 0, len = categories.length, category, marginLeft = 48, listItem;
+        var list = $('<ul />', {'data-cid': parentId});
+
+		if(level > 0){
+			list.css('margin-left', marginLeft);
+		}
+
+        for(i; i < len; ++i){
+            category = categories[i];
+
+             listItem = $('<li></li>', {'data-cid': category.cid})
+                 .append(renderListItem(category))
+                 .appendTo(list);
+
+            if(category.disabled){
+                listItem.addClass('disabled');
+            }
+
+            if(category.children.length > 0){
+                renderList(category.children, level + 1, listItem, category.cid);
+            }
+        }
+
+        list.appendTo(container);
+		sortables[parentId] = Sortable.create(list[0], {
+			group: 'cross-categories',
+			animation: 150,
+			handle: '.icon',
+            dataIdAttr: 'data-cid',
+			ghostClass: "placeholder",
+			onAdd: itemDidAdd,
+			onEnd: itemDragDidEnd
+		});
+    }
+
+    function renderListItem(categoryEntity){
+        var listItem = $(templates.parse(
+            '<div class="row">' +
```

Comment:
Yes, It will be much better to have template in external file. I didn't know about such option.
Does `template.js` use cache for template fetching? Is there any type of pre-fetching of template?
My concern, will it work Ok in iteration? I mean, if this solution produce 20 HTTP requests for list of 20 items... current solution will be better still.